Mathlib/Algebra/Lie/Rank.lean

```diff
@@ -0,0 +1,251 @@
+if the `n`-th coefficient of the characteristic polynomial of `ad R L x` is non-zero.
+
+## Main declarations
+
+* `LieAlgebra.rank R L` is the rank of a Lie algebra `L` over a commutative ring `R`.
+* `LieAlgebra.IsRegular R x` is the predicate that an element `x` of a Lie algebra `L` is regular.
+
+## References
+
+* [barnes1967]: "On Cartan subalgebras of Lie algebras" by D.W. Barnes.
+
+-/
+
+open scoped BigOperators
+
+variable {R L M ι ι' ιM ιM' : Type*}
+variable [CommRing R] [LieRing L] [LieAlgebra R L]
+variable [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+variable [Fintype ι] [Fintype ιM] [Fintype ι'] [Fintype ιM']
+
+namespace LieAlgebra
+
+local notation "φ" => LieModule.toEndomorphism R L M
+
+section basic
+
+variable [DecidableEq ιM] [DecidableEq ιM']
+variable (b : Basis ι R L) (bₘ : Basis ιM R M) (b' : Basis ι' R L) (bₘ' : Basis ιM' R M)
+
+open LieModule LinearMap MvPolynomial in
+/-- Let `M` be a Lie module of a Lie algebra `L` over `R`,
+and let `b` be a basis of `L` and `bₘ` a basis of `M`.
+Then `lieMatrixPoly b bₘ (i,j)` is the polynomial that evaluates on elements `x` of `L`
+to the `(i,j)`-th coefficient of the matrix representation of `⁅x, ·⁆` acting on `M`. -/
+noncomputable
+def lieMatrixPoly (ij : ιM × ιM) : MvPolynomial ι R :=
+  ∑ k : ι, monomial (.single k 1) ((toMatrix bₘ bₘ <| φ <| b k) ij.1 ij.2)
+
+open LinearMap in
+@[simp]
+lemma lieMatrixPoly_eval_eq_toMatrix (ij : ιM × ιM) (c : ι →₀ R) :
+    MvPolynomial.eval c (lieMatrixPoly b bₘ ij) =
+      (toMatrix bₘ bₘ <| φ (b.repr.symm c)) ij.1 ij.2 := by
+  rcases ij with ⟨i, j⟩
+  simp only [lieMatrixPoly, map_sum, Basis.repr_symm_apply, MvPolynomial.eval_monomial,
+    pow_zero, Finsupp.prod_single_index, pow_one]
+  induction c using Finsupp.induction_linear
+  case h0 => simp
+  case hadd f g hf hg => simp [hf, hg, mul_add, Finset.sum_add_distrib]
+  case hsingle k r =>
+    rw [Finset.sum_eq_single k, Finsupp.single_eq_same]
+    · rw [mul_comm, Finsupp.total_single, LieHom.map_smul, map_smul, Matrix.smul_apply, smul_eq_mul]
+    · rintro l - hl; rw [Finsupp.single_eq_of_ne hl.symm, mul_zero]
+    · simp only [Finset.mem_univ, not_true_eq_false, IsEmpty.forall_iff]
+
+lemma lieMatrixPoly_eval_eq_lie (ij : ιM × ιM) (c : ι →₀ R) :
+    MvPolynomial.eval c (lieMatrixPoly b bₘ ij) = bₘ.repr ⁅b.repr.symm c, bₘ ij.2⁆ ij.1 := by
+  rw [lieMatrixPoly_eval_eq_toMatrix, Basis.repr_symm_apply, LinearMap.toMatrix_apply,
+    LieModule.toEndomorphism_apply_apply]
+
+open Matrix
+
+/-- Let `M` be a Lie module of a Lie algebra `L` over `R`,
+and let `b` be a basis of `L` and `bₘ` a basis of `M`.
+Then `lieCharpoly b bₘ` is the polynomial that evaluates on elements `x` of `L`
+to the characteristic polynomial of `⁅x, ·⁆` acting on `M`. -/
+noncomputable
+def lieCharpoly : Polynomial (MvPolynomial ι R) :=
+  (charpoly.univ R ιM).map <| MvPolynomial.bind₁ (lieMatrixPoly b bₘ)
```

Comment:
It looks to me like this definition has an artificial depenence on `bₘ` (in the sense that the choice doesn't matter). Do I understand correctly that we should instead use:
```suggestion
variable (M) in
/-- Let `M` be a Lie module of a Lie algebra `L` over `R`,
and let `b` be a basis of `L` and `bₘ` a basis of `M`.
Then `lieCharpoly b bₘ` is the polynomial that evaluates on elements `x` of `L`
to the characteristic polynomial of `⁅x, ·⁆` acting on `M`. -/
noncomputable
def lieCharpoly [Module.Free R M] [Module.Finite R M] : Polynomial (MvPolynomial ι R) :=
  (charpoly.univ R (Module.Free.ChooseBasisIndex R M)).map <|
    MvPolynomial.bind₁ (lieMatrixPoly b (Module.Free.chooseBasis R M))
```

Reply:
We could indeed do that. But I wanted to keep the basis flexible, so that it will be easier to show at some point that the result doesn't depend on the choice of basis.

Reply:
I appreciate your point but I'm not (yet?) convinced. I note that elsewhere in the library we follow the pattern of making an arbitrary choice. For example `LinearMap.trace` or `LinearMap.det`.

Reply:
But they come with lemmas that show you can use an arbitrary basis to convert to the corresponding concept for matrices. Because we want polynomials everywhere, we can't even really state such lemmas, unless we allow for the flexibility of choosing a basis in these decls.

Reply:
This has been factored out into a new theory file. So this is resolved.